xmppserver/src/main/java/org/jivesoftware/openfire/net/ServerStanzaHandler.java

```diff
@@ -132,6 +150,48 @@ void startTLS() throws Exception {
         //needed ? Connection.ClientAuth.needed : Connection.ClientAuth.wanted
         connection.startTLS(false, false);
     }
+
+    @Override
+    protected String getStreamHeader() {
+        StringBuilder sb = new StringBuilder(200);
+        sb.append("<?xml version='1.0' encoding='");
+        sb.append(CHARSET);
+        sb.append("'?>");
+        sb.append("<stream:stream xmlns:stream=\"http://etherx.jabber.org/streams\" xmlns=\"");
+        sb.append(getNamespace()).append("\"");
+        sb.append(getAdditionalNamespaces());
+        sb.append(" from=\"").append(domainPair.getLocal()).append("\"");
+        sb.append(" to=\"").append(domainPair.getRemote()).append("\"");
+        sb.append(" id=\"").append(session.getStreamID()).append("\"");
+        sb.append(" xml:lang=\"").append(session.getLanguage().toLanguageTag()).append("\"");
+        sb.append(" version=\"").append(Session.MAJOR_VERSION).append('.').append(Session.MINOR_VERSION);
+        sb.append("\">");
+        return sb.toString();
+}
+
+    @Override
+    protected void tlsNegotiated(XmlPullParser xpp) throws XmlPullParserException, IOException {
+        // Discard domain values obtained from unencrypted initiating stream
```

Comment:
This code is server-to-server specific (with usage of DomainPair and ServerSession-usage), but the specification in the RFC that you're referncing hold true for all TLS negotations. Should something similar be built for the other StanzaHandlers?

Reply:
Possibly, but the ticket scope is for S2S issues.

Reply:
Fair. I've raised https://igniterealtime.atlassian.net/browse/OF-2645 to track this as a future improvement.